// space-lift/commonjs/wrapper.js
"use strict";
exports.__esModule = true;
var lift = function (obj) {
    if (obj instanceof Array)
        return new ArrayOps(obj);
    if (typeof obj === 'string')
        return new StringOps(obj);
    if (typeof obj === 'number')
        return new NumberOps(obj);
    if (obj === true || obj === false)
        return new BoolOps(obj);
    return new ObjectOps(obj);
};
exports["default"] = lift;
function getValue(input) {
    return input && input['_isLiftWrapper']
        ? input.value()
        : input;
}
exports.getValue = getValue;
//--------------------------------------
//  Array
//--------------------------------------
var ArrayOps = (function () {
    function ArrayOps(array) {
        this._isLiftWrapper = true;
        this._value = array;
    }
    ArrayOps.prototype.value = function () { return this._value; };
    return ArrayOps;
}());
exports.ArrayOps = ArrayOps;
//--------------------------------------
//  Object
//--------------------------------------
var ObjectOps = (function () {
    function ObjectOps(object) {
        this._isLiftWrapper = true;
        this._value = object;
    }
    ObjectOps.prototype.value = function () { return this._value; };
    return ObjectOps;
}());
exports.ObjectOps = ObjectOps;
//--------------------------------------
//  Number
//--------------------------------------
var NumberOps = (function () {
    function NumberOps(num) {
        this._isLiftWrapper = true;
        this._value = num;
    }
    NumberOps.prototype.value = function () { return this._value; };
    return NumberOps;
}());
exports.NumberOps = NumberOps;
//--------------------------------------
//  String
//--------------------------------------
var StringOps = (function () {
    function StringOps(str) {
        this._isLiftWrapper = true;
        this._value = str;
    }
    StringOps.prototype.value = function () { return this._value; };
    return StringOps;
}());
exports.StringOps = StringOps;
//--------------------------------------
//  Boolean
//--------------------------------------
// Not that we expect to expand on the boolean capabilities... But for completeness sake.
var BoolOps = (function () {
    function BoolOps(value) {
        this._isLiftWrapper = true;
        this._value = value;
    }
    BoolOps.prototype.value = function () { return this._value; };
    return BoolOps;
}());
exports.BoolOps = BoolOps;
